keps/sig-apps/85-Graduate-PDB-to-Stable/README.md

```diff
@@ -0,0 +1,319 @@
+
+### API changes
+
+* Add conditions to the status object for PodDisruptionBudget.
+* Change the semantics of an empty selector.
+
+
+### Test Plan
+
+#### Existing Tests
+PodDisruptionBudget currently has tests in various components that use the feature:
+
+* Unit tests for disruption controller
+  - https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/disruption/disruption_test.go
+* Integration tests for disruption controller
+  - https://github.com/kubernetes/kubernetes/blob/master/test/integration/disruption/disruption_test.go
+* Kubectl
+  - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/generate/versioned/pdb_test.go
+  tests generation of a PDB objects out of given parameters
+  - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/create/create_pdb_test.go
+  tests creation of PDB objects from cmd parameters
+* Scheduler
+  - https://github.com/kubernetes/kubernetes/blob/4b59044b8d2a3502ea490ba2c958008a098511a3/test/integration/scheduler/preemption_test.go#L1053
+  tests effects of PDB on preemption (PDB is honored in a best effort way)
+* Eviction integration tests
+  - https://github.com/kubernetes/kubernetes/blob/master/test/integration/evictions/evictions_test.go test eviction logic and its interactions with PDB.
+* Autoscaler
+  - https://github.com/kubernetes/kubernetes/blob/master/test/e2e/autoscaling/cluster_size_autoscaling.go ensure that Autoscaler respects PDB when draining nodes.
+* Integration tests for disruption controller
+  - https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apps/disruption.go
+
+All beta PDB endpoints are covered by e2e tests: https://apisnoop.cncf.io/1.20.0/beta/policy
+
+#### Conformance tests
+
+The following e2e tests will be included in the conformance tests:
+
+ * DisruptionController
+   * should block an eviction until the PDB is updated to allow it
+   * create a PodDisruptionBudget
+   * delete collection of PodDisruptionBudget
+   * delete a PodDisruptionBudget
+   * list or watch objects of kind PodDisruptionBudget
+   * partially update the specified PodDisruptionBudget
+   * partially update status of the specified PodDisruptionBudget
+   * read the specified PodDisruptionBudget
+   * read status of the specified PodDisruptionBudget
+   * replace the specified PodDisruptionBudget
+   * replace status of the specified PodDisruptionBudget
+
+### Graduation Criteria
+
+Graduation to GA:
+- [x] Implement Mutable PDBs
+- [ ] Address performance issues
+- [ ] Pass conformance tests
+- [ ] Update documents to reflect the changes
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: PodDisruptionBudget
+    - Components depending on the feature gate: kube-controller-manager
+
```

Comment:
Shouldn't we also use this feature-gate in kube-apiserver?

i.e. when PDB feature is switched off, eviction subresource doesn't look into PDBs...

@liggitt - for thoughts

Reply:
Yeah, I think we should make sure that the eviction functionality is also aware of the PodDisruptionBudget feature gate and doesn't look at PDBs if the feature is not enabled.

Reply:
I'm a bit confused how this helps advance PDB to GA... once GA, this feature gate would be locked on. Do we expect/want clusters to use the eviction subresource without PDBs being involved?

Reply:
Yeah - it's a bit too late, I was wondering if we should consider backporting adding the featue-gate in apiserver, for the purpose:
- I'm enabling beta APIs
- I'm upgrading my cluster to 1.21 where PDB will get GA
- I created some PDBs and start using them
- I'm downgrading back to 1.20 (because of whatever reason)
- and now I'm potentially stuck, because I can't evict anything, because nothing is updating AllowedDisruptions in my PDB objects...

Reply:
I wouldn't start allowing the eviction subresource to be used without PDB at this point. If you don't have the PDB API enabled today, calls to eviction fail (because the PDB API 404's)

Reply:
Sorry - that's not what I had on my mind.
Let' me be more specific:

1. Let's say I have PDB API already enabled and will not disable it later too.
[The feature gate doesn't control if the API is disabled or not]
2. I'm using 1.20 where I have: (a) Beta APIs enabled, (b) Beta features disabled
3. I'm upgrading to 1.21, created PDB and started using it
4. I'm downgrading back to 1.20, so PDB API remains enabled, but feature gets disabled (which stops updating my PDB object)
5. and now I'm potentially stuck, because I can't evict anything, because nothing is updating AllowedDisruptions in my PDB objects...

So it's not about allowing using eviction without PDB.

I think that requiring PDB API to be abled as is now is completely fine.
But what if I have the API enabled, but the feature itself is disabled?

[it basically relates to what is described in the answer for a question below:
```
  Yes, but any existing PDBs will no longer be updated and the old state might
  still block eviction of pods. Deleting a PDB from the cluster will remove any
  restrictions on evictions for the pods covered by the selector.
```

I'm asking whether this is what we really want or do we want to protect from that somehow?

@liggitt ^^

Reply:
So if we wanted to have the eviction API aware of the PDB feature gate, then we need to either backport it to the 1.20 release and then rely on people to update to the patch version before going to 1.21, or we would have to put that in 1.21 and then delay GA to 1.22 right?

Based on the comments here, it sounds like we probably are ok moving forward without doing this. Is that ok with you @wojtek-t ?

Reply:
Yeah - I think we're fine with just moving as is with this one. It's not ideal, but as Jordan wrote we're in that state for long time and it's also documented in the PRR below.